### DISCUSSION TITLE
Async Generator without yield typed as Coroutine

### DISCUSSION THREAD

**Discussion by aidandj** [microsoft/pyright#7128] [General]
When a coroutine is defined as returning an async generator, if there is no yield statement it is typed as a coroutine.

I see the same behavior in mypy, but can't tell if it is a bug in both, or by design.

```python
import typing as t

async def generator_without_yield() -> t.AsyncGenerator[str, None]:
    raise NotImplementedError()

reveal_type(generator_without_yield)

async def generator_with_yield() -> t.AsyncGenerator[str, None]:
    yield "foo"

reveal_type(generator_with_yield)
```

**Comment by erictraut** [comment]
I don't think this is a bug. The Python compiler creates a `Coroutine` object in the first case. If your coroutine implementation constructs and return an `AsyncGenerator` object, then it will be wrapped in a `Coroutine` object, hence the type `Coroutine[Any, Any, AsyncGenerator[str, None]]`.

You can see this by adding the lines:

```python
print(generator_without_yield())
print(generator_with_yield())
```

If you then run the code, you'll see the following output:
```
<coroutine object generator_without_yield at 0x104f22f00>
<async_generator object generator_with_yield at 0x100b62f00>
```

Pyright and mypy are simply modeling the runtime behavior here.
